core/services/relay/evm/mercury/v1/report_codec.go

```diff
@@ -0,0 +1,119 @@
+package mercury_v1
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/pkg/errors"
+	ocrtypes "github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+
+	relaymercury "github.com/smartcontractkit/chainlink-relay/pkg/reportingplugins/mercury"
+	reportcodec "github.com/smartcontractkit/chainlink-relay/pkg/reportingplugins/mercury/v1"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/mercury/types"
+)
+
+var ReportTypes = getReportTypes()
+
+func getReportTypes() abi.Arguments {
+	mustNewType := func(t string) abi.Type {
+		result, err := abi.NewType(t, "", []abi.ArgumentMarshaling{})
+		if err != nil {
+			panic(fmt.Sprintf("Unexpected error during abi.NewType: %s", err))
+		}
+		return result
+	}
+	return abi.Arguments([]abi.Argument{
+		{Name: "feedId", Type: mustNewType("bytes32")},
+		{Name: "observationsTimestamp", Type: mustNewType("uint32")},
+		{Name: "validFromTimestamp", Type: mustNewType("uint32")},
+		{Name: "benchmarkPrice", Type: mustNewType("int192")},
+		{Name: "bid", Type: mustNewType("int192")},
+		{Name: "ask", Type: mustNewType("int192")},
+		{Name: "expiresAt", Type: mustNewType("uint32")},
+		{Name: "linkFee", Type: mustNewType("int192")},
+		{Name: "nativeFee", Type: mustNewType("int192")},
+	})
+}
+
+var _ reportcodec.ReportCodec = &ReportCodec{}
+
+type ReportCodec struct {
+	logger logger.Logger
+	feedID types.FeedID
+}
+
+func NewReportCodec(feedID [32]byte, lggr logger.Logger) *ReportCodec {
+	return &ReportCodec{lggr, feedID}
+}
+
+func (r *ReportCodec) BuildReport(paos []reportcodec.IParsedAttributedObservation, f int, validFromTimestamp, expiresAt uint32) (ocrtypes.Report, error) {
+	if len(paos) == 0 {
+		return nil, errors.Errorf("cannot build report from empty attributed observations")
+	}
+
+	mPaos := reportcodec.Convert(paos)
+
+	timestamp := relaymercury.GetConsensusTimestamp(mPaos)
+
+	// todo: add checks for validFromTimestamp
+
+	benchmarkPrice, err := relaymercury.GetConsensusBenchmarkPrice(mPaos, f)
+	if err != nil {
+		return nil, errors.Wrap(err, "GetConsensusBenchmarkPrice failed")
+	}
+	bid, err := relaymercury.GetConsensusBid(mPaos, f)
+	if err != nil {
+		return nil, errors.Wrap(err, "GetConsensusBid failed")
+	}
+	ask, err := relaymercury.GetConsensusAsk(mPaos, f)
+	if err != nil {
+		return nil, errors.Wrap(err, "GetConsensusAsk failed")
+	}
+	linkFee, err := relaymercury.GetConsensusLinkFee(mPaos, f)
+	if err != nil {
+		return nil, errors.Wrap(err, "GetConsensusLinkFee failed")
+	}
+	nativeFee, err := relaymercury.GetConsensusNativeFee(mPaos, f)
+	if err != nil {
+		return nil, errors.Wrap(err, "GetConsensusNativeFee failed")
+	}
```

Comment:
To be honest all these consensus method calls could (should?) be moved to the relay and we can simply pass the args to `BuildReport` directly, making it much simpler

Reply:
This way `BuildReport` has some flexibility in changing the report structure built from paos without any signature changes. I'm actually wondering if this whole thing should be in the relay - the functionality separation between core and relay is somewhat unclear to me.

Reply:
Chain-specific goes in core, chain-agnostic goes in relay